synapse/storage/registration.py

```diff
@@ -984,3 +983,191 @@ def _bg_user_threepids_grandfather_txn(txn):
+        client_secret/medium/address combo
+
+        Args:
+            medium (str): The medium of the 3PID
+            address (str): The address of the 3PID
+            client_secret (str): A unique string provided by the client to
+                help identify this validation attempt
+
+        Returns:
+            deferred {str, int}|None: A dict containing the
+                latest session_id and send_attempt count for this 3PID.
+                Otherwise None if there hasn't been a previous attempt
+        """
+        row = yield self._simple_select_one(
+            table="threepid_validation_session",
+            keyvalues={
+                "medium": medium,
+                "address": address,
+                "client_secret": client_secret,
+            },
+            retcols=["session_id", "last_send_attempt"],
+            allow_none=True,
+            desc="get_threepid_validation_session",
+        )
+
+        defer.returnValue(row)
+
+    @defer.inlineCallbacks
+    def validate_threepid_validation_token(
+        self,
+        session_id,
+        client_secret,
+        token,
+        current_ts,
+    ):
+        """Validate a validation token, session_id and client_secret
+
+        Args:
+            session_id (str): The id of a validation session
+            client_secret (str): A unique string provided by the client to
+                help identify this validation attempt
+            token (str): A validation token
+            current_ts (int): The current unix time in milliseconds. Used for
+                checking token expiry status
+
+        Returns:
+            deferred (bool, str|None): A tuple where the first value represents
+                whether the provided sid/client_secret/token combo was correct,
+                and the second being the link to redirect the user to if there
+                is one
+        """
+        # Check that everything matches in the DB
+        def validate_threepid_validation_token_txn(
+            txn,
+            token,
+            client_secret,
+            ts,
+        ):
+            sql = ("SELECT token, next_link FROM threepid_validation_token "
+                   "WHERE token = ? AND session_id = (SELECT session_id FROM "
+                   "threepid_validation_session WHERE client_secret = ?) AND "
+                   "expires > ? LIMIT 1")
+            txn.execute(sql, (token, client_secret, ts))
+
+            row = txn.fetchone()
+
+            # Return None if we didn't get a match
+            if not row:
+                return None, None
+
+            return row[0], row[1]
+
+        a_token, next_link = yield self.runInteraction(
+            "validate_threepid_validation_token",
+            validate_threepid_validation_token_txn,
+            token,
+            client_secret,
+            current_ts,
+        )
+
+        # Check if we got a match
+        defer.returnValue((a_token is not None, next_link))
```

Comment:
I'd make the inner function return the correct thing, and then get rid of the inlineCallbacks and return the result of `runInteraction` directly.